### PR TITLE
Don't reuse GCS BatchRequest objects CROM-6708 CROM-6709

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala:71]:
```suggestion
    * The BatchRequest's internal queue is NOT cleared when exceptions are thrown inside BatchRequest.execute().
```

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala:72]:
```suggestion
    * Any subsequent enqueueing operations only add to the internal queue and do not replace the internal queue.
```

**Review by aednichols — COMMENTED** [review]
Nice find, reminds me of https://github.com/broadinstitute/firecloud-orchestration/pull/665

One thing I don't quite understand is why the [batch failure log message](https://github.com/broadinstitute/cromwell/blob/e9384f63704fafb524371415fd4c1faf96023b90/engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala#L139) topped out at 100 requests (as designed) according to the logs. Are there two different levels of batching at play here?

**Comment by kshakir** [comment]
The messages are logging the size of the list being (re-)added to the `BatchRequest`, not what's inside the possibly stale `ArrayList` inside the `BatchRequest` object.

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala:74]: For the sanity of future maintainers it might be worth mentioning your finding that the "Received" number in the error message always seems to be 3000 even when the batch was known to have > 3000 elements:

```
413 Request Entity Too Large
POST https://storage.googleapis.com/batch/storage/v1
Inner request count exceeds the limit. Received: 3000, Limit: 3000
```

**Review by mcovarr — APPROVED** [review]
if Travis / Circle are ok with it 👍

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala
Comment (aednichols) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala:167]:
```suggestion
        logger.info(s"Failed to execute GCS Batch request. Failed request belonged to batch of size ${batchRequest.size()} containing commands: " +
```

**Comment by mcovarr** [comment]
> The messages are logging the size of the list being (re-)added to the BatchRequest, not what's inside the possibly stale ArrayList inside the BatchRequest object.

Yeah okay maybe don't mention that then since it will force those future maintainers to imagine what was happening before this variable became a local...

File: engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala
Comment (aednichols) [inline engine/src/main/scala/cromwell/engine/io/gcs/GcsBatchFlow.scala:171]:
```suggestion
        logger.info(s"Failed to execute GCS Batch request. Failed request belonged to batch of size ${batchRequest.size()} containing commands: " +
```

**Review by aednichols — APPROVED** [review]
Pretty sure I added this logging originally and its inaccuracy definitely made things harder to debug, so here's an improvement

**Review by aednichols — COMMENTED** [review]
Sorry to pile on with yet another review... I feel like this could use a changelog so that we can easily identify which version fixed the bug when (inevitably) we want to know in the future

> Fixed a bug that could cause workflows to unexpectedly fail with errors related to Google Cloud Storage. The errors reference `GcsBatchFlow.BatchFailedException` and either `Read timed out` or `413 Request Entity Too Large`.

**Comment by kshakir** [comment]
No worries, that's what PRs are for. 😉 

Regarding the text:

> Fixed a bug that could cause workflows to unexpectedly fail with errors related to Google Cloud Storage. The errors reference GcsBatchFlow.BatchFailedException and Read timed out or 413 Request Entity Too Large.

This PR doesn't address the _initial_ read timeouts. But it does mitigate them piling up to create bigger batches that might cause more timeouts.

What I'm not sure of is with this change of a new-batch-per-request will the "Attempted 10 times" be able to weather whatever is causing the timeout hiccups or not.

I don't want to overpromise in the message, so maybe something like?

> Better handling for a bug that could cause workflows to unexpectedly fail with errors related to Google Cloud Storage. The errors reference GcsBatchFlow.BatchFailedException and Read timed out or 413 Request Entity Too Large.

**Comment by aednichols** [comment]
Ah yes, let's just claim we fixed one bug then 😄 
My take; use, modify, or discard as desired:
> Fixed a bug that could cause workflows to unexpectedly fail with the error `413 Request Entity Too Large` when accessing Google Cloud Storage.

**Comment by kshakir** [comment]
Cool. 59 is scheduled to go out before March 26. I'll edit the changelog once I figure out if the lab wants to:
- wait for us to finish fixing our CI and patch 58, or
- take a 59-SNAP and this fix may not be hotfixed

**Comment by kshakir** [comment]
Changelog updated and this PR is available for re-review.

The lab deployed a `-SNAP` version. Updated the changelog list this change as part of `59` and filed BT-187 regarding hotfix PRs breaking in Circle CI.

**Review by mcovarr — APPROVED** [review]
Thank you for the addition and fix to the changelog!

**Pull request merged**
